firedrake/preconditioners/facet_split.py

```diff
@@ -0,0 +1,252 @@
+from firedrake.petsc import PETSc
+from firedrake.preconditioners.base import PCBase
+import firedrake.dmhooks as dmhooks
+import numpy
+
+
+__all__ = ['FacetSplitPC']
+
+
+def get_permutation_map(V, W):
+    from firedrake.preconditioners.fdm import glonum_fun, restricted_dofs
+    bsize = V.value_size
+    V_indices = None
+    V_map, nel = glonum_fun(V.cell_node_map(), bsize=bsize)
+    perm = numpy.empty((V.dof_count, ), dtype=PETSc.IntType)
+    perm.fill(-1)
+
+    offset = 0
+    for Wsub in W:
+        W_indices = None
+        W_map, nel = glonum_fun(Wsub.cell_node_map(), bsize=bsize)
+        rdofs = restricted_dofs(Wsub.finat_element, V.finat_element)
+
+        for e in range(nel):
```

Comment:
Element loop in Python code is usually a problem. Are we really sure this won't be expensive for large meshes? Should it not be in cython?